### PR TITLE
page_id: KeyPath to PageId Iterator

### PR DESCRIPTION
**Pull request by gabriele-0201**

### CODE REVIEW

**Comment by gabriele-0201** [comment]
* **#28** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/28?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: core/src/pageId.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        let sextet = self
+            .key_path
+            .view_bits::<Msb0>()
+            .chunks_exact(DEPTH)
+            .nth(self.level as usize)?;
+
+        //#[cfg(feature = "std")]
+        //println!("level: {}, sextet: {:?}", self.level, sextet);
+
+        let next_page_id: &mut BitSlice<u8, Msb0> = self.next_page_id.view_bits_mut();
```
Comment (gabriele-0201) [inline core/src/pageId.rs:59]: Using Msb0 for the KeyPath is perfect, to maintain consistency I tried to use the same convention for the PageId as well. It's feasible but the indices used are not very convenient in this case, especially when working mainly with the least significant byte. Using an Lsb0 convention could simplify things.

The handling of PageId is mainly restricted to this Iterator, so there are no other places where one must remember that the PageId is a little-endian integer. What do you think? @rphmeier @pepyakin should we keep Msb0 also here?

Comment (rphmeier) [reply core/src/pageId.rs:59]: How does Lsb0 make it easier?

File: core/src/pageId.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    fn next(&mut self) -> Option<Self::Item> {
+        let prev_page_id = self.next_page_id;
+
+        // The last bits will not be used as a sextet to identify a page.
+        // If sextets are finished, then there are no more pages to load.
+        let sextet = self
+            .key_path
+            .view_bits::<Msb0>()
+            .chunks_exact(DEPTH)
+            .nth(self.level as usize)?;
```
Comment (rphmeier) [inline core/src/pageId.rs:54]: isn't this causing `O(n^2)` behavior?

Comment (rphmeier) [reply core/src/pageId.rs:54]: i recommend keeping either an iterator or an index inside of the `PageIdsIterator` struct so it's not being recomputed each time

File: core/src/lib.rs [hunk 3]
```diff
@@ -10,5 +10,7 @@
 
 extern crate alloc;
 
+pub mod page;
+pub mod pageId;
```
Comment (rphmeier) [inline core/src/lib.rs:14]: recommend combining the page / page ID into one `page` module. if not, this should be `page_id` - Rust is always `snake_case`

**Review by rphmeier — COMMENTED** [review]
I'm somewhat tempted to say "use a wide int library" for this binary logic, but it may be overkill, unless there's a very light one, given that you've already implemented the logic by hand.

File: core/src/pageId.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+///
+/// Given a KeyPath, it will be divided into chunks of 6 bits (sextet), forming
+/// an array of sextets. Each PageId will be constructed in the following manner:
+///
+/// page_ids[i] = (prev_page_id << d) + int(sextets[i]) + 1
+///
+/// The following struct implements an Iterator of PageIds over a KeyPath
+/// to lazily construct them as needed
+struct PageIdsIterator {
+    key_path: KeyPath,
```
Comment (rphmeier) [inline core/src/pageId.rs:28]: does the iterator need to own the path of just borrow it?

File: core/src/page.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+//! TODO
+
+/// Depth of the rootless sub-binary tree stored in a page
+pub const DEPTH: usize = 6;
+
+// Total number of nodes stored in one Page. It depends on the `DEPTH`
+// of the rootless sub-binary tree stored in a page following this formula:
+// (2^DEPTH) - 2
+pub const NODES_PER_PAGE: usize = 126;
```
Comment (rphmeier) [inline core/src/page.rs:9]: could also be specified as `(1 << DEPTH) - 2`

File: core/src/pageId.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            .view_bits::<Msb0>()
+            .chunks_exact(DEPTH)
+            .nth(self.level as usize)?;
+
+        //#[cfg(feature = "std")]
+        //println!("level: {}, sextet: {:?}", self.level, sextet);
+
+        let next_page_id: &mut BitSlice<u8, Msb0> = self.next_page_id.view_bits_mut();
+
+        if sextet != bits!(1, 1, 1, 1, 1, 1) {
```
Comment (rphmeier) [inline core/src/pageId.rs:61]: I recommend extracting out a few standalone functions:

```rust
pub fn child_page_id(page_id, child_index) -> PageId;
pub fn parent_page_id(page_id) -> PageId;
```

and then implementing this iterator based on `child_page_id`.

Comment (rphmeier) [reply core/src/pageId.rs:61]: https://github.com/thrumdev/nomt/pull/29 needs these

File: core/src/pageId.rs (hunk 2 above)
Comment (gabriele-0201) [reply core/src/pageId.rs:54]: Yes! keeping a reference to the key path and iterating it when needed is the idea, I just made the easiest solution to begin with to focus on the PageIds generation. I will implement it tomorrow!

File: core/src/lib.rs (hunk 3 above)
Comment (gabriele-0201) [reply core/src/lib.rs:14]: Oh, yes, I'm stupid 😅

File: core/src/pageId.rs (hunk 4 above)
Comment (gabriele-0201) [reply core/src/pageId.rs]: Currently, it borrows it

File: core/src/page_id.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+impl<'a> Iterator for PageIdsIterator<'a> {
+    type Item = PageId;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let new_page_id = match self.page_id {
+            None => Some(ROOT_PAGE_ID),
+            Some(prev_page_id) => {
+                // If sextets are finished, then there are no more pages to load
+                let child_index = self.sextets.next()?.load_be::<u8>();
+                // TODO: could be nice to modify prev_page_id in place
```
Comment (rphmeier) [inline core/src/page_id.rs:96]: let's remove this TODO in a follow up. Very premature optimization

File: core/src/page_id.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+///
+/// The following struct implements an Iterator of PageIds over a KeyPath
+/// to lazily construct them as needed
+struct PageIdsIterator<'a> {
+    sextets: ChunksExact<'a, u8, Msb0>,
+    page_id: Option<PageId>,
+}
+
+impl<'a> PageIdsIterator<'a> {
+    fn new(key_path: &'a KeyPath) -> Self {
```
Comment (rphmeier) [inline core/src/page_id.rs:78]: unused - should be ignored or made public

File: core/src/page_id.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+/// bytes will be treated as a 256-bit big-endian integer.
+///
+/// Given a KeyPath, it will be divided into chunks of 6 bits (sextet), forming
+/// an array of sextets. Each PageId will be constructed in the following manner:
+///
+/// page_ids[i] = (prev_page_id << d) + int(sextets[i]) + 1
+///
+/// The following struct implements an Iterator of PageIds over a KeyPath
+/// to lazily construct them as needed
+struct PageIdsIterator<'a> {
```
Comment (rphmeier) [inline core/src/page_id.rs:72]: pub

File: core/src/page_id.rs [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+const MAX_CHILD_INDEX: u8 = (1 << DEPTH) - 1;
+
+pub fn child_page_id(mut page_id: PageId, child_index: u8) -> PageId {
+    // TODO: proper error handling
+    assert!(child_index & 0b11000000 == 0);
+
+    let page_id_bits = page_id.view_bits_mut::<Msb0>();
+
+    if child_index ^ MAX_CHILD_INDEX == 0 {
```
Comment (rphmeier) [inline core/src/page_id.rs:18]: why not just use `==`? seems unnecessarily fancy

File: core/src/page_id.rs [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+pub type PageId = [u8; 32];
+
+pub const ROOT_PAGE_ID: [u8; 32] = [0; 32];
+
+const MAX_CHILD_INDEX: u8 = (1 << DEPTH) - 1;
+
+pub fn child_page_id(mut page_id: PageId, child_index: u8) -> PageId {
+    // TODO: proper error handling
+    assert!(child_index & 0b11000000 == 0);
```
Comment (rphmeier) [inline core/src/page_id.rs:14]: this is less readable than `assert!(child_index < MAX_CHILD_INDEX)`. We should not be making the reader spend cycles on things that can be easy unless there is a driving performance need.

Comment (rphmeier) [reply core/src/page_id.rs:14]: also, generally I consider `assert` or panicking bad practice over error handling, but there is a TODO which should include the two failure cases:
  1. child index out of bounds
  2. page id overflow

File: core/src/page.rs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+//! TODO: add docs
+
+/// Depth of the rootless sub-binary tree stored in a page
+pub const DEPTH: usize = 6;
+
+// Total number of nodes stored in one Page. It depends on the `DEPTH`
+// of the rootless sub-binary tree stored in a page
+pub const NODES_PER_PAGE: usize = (1 << DEPTH) - 2;
```
Comment (rphmeier) [inline core/src/page.rs:8]: (1 << DEPTH + 1) - 2 actually (my bad earlier)

File: core/src/page_id.rs [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+                Some(child_page_id(prev_page_id, child_index))
+            }
+        };
+        self.page_id = new_page_id;
+        new_page_id
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (rphmeier) [inline core/src/page_id.rs:107]: let's add more comprehensive tests including the overflow / out of bounds cases

**Review by rphmeier — COMMENTED** [review]
documentation on all public items is missing, let's add that & address other comments in a follow up.

File: core/src/page_id.rs (hunk 10 above)
Comment (gabriele-0201) [reply core/src/page_id.rs:18]: Totally true, I started thinking about using a mask and I went in that direction but yes,`==` does the same thing and it is cleaner

File: core/src/page_id.rs (hunk 11 above)
Comment (gabriele-0201) [reply core/src/page_id.rs:14]: thanks, yes `assert!(child_index <= MAX_CHILD_INDEX)` is for sure more readable. I will implement proper error handling though!
